synapse/app/generic_worker.py

```diff
@@ -206,10 +206,16 @@ async def on_POST(self, request, device_id):
 
         if body:
             # They're actually trying to upload something, proxy to main synapse.
-            # Pass through the auth headers, if any, in case the access token
-            # is there.
-            auth_headers = request.requestHeaders.getRawHeaders(b"Authorization", [])
-            headers = {"Authorization": auth_headers}
+
+            # Proxy headers from the original request, such as the auth headers
+            # (in case the access token is there) and the original IP /
+            # User-Agent of the request.
+            headers = {}
+            for header in (b"Authorization", b"X-Forwarded-For", b"User-Agent"):
+                header_value = request.requestHeaders.getRawHeaders(header)
+                if header_value:
+                    headers[header] = header_value
```

Comment:
This is a slight behavior change for the `Authorization` header, which was previously always provided, but sometimes was an empty list. I think it ends up being the same behavior though since Twisted essentially just ignored it if there was no values:

https://github.com/twisted/twisted/blob/twisted-20.3.0/src/twisted/web/http.py#L1152-L1154

We can't use that code for the `User-Agent` header though since we only want to override it if one exists on the incoming request. (Or maybe we also want to override it? Not sure.)

Reply:
I think we always want to override the user agent, as other we'll persist the wrong user agent to the DB?

Reply:
Alright, so in the case of no user-agent being on the original request the proxied request would also have no user-agent.

Reply:
I made this change, see 17b916db475708f5d2751d7f17d8380ac9317c61.